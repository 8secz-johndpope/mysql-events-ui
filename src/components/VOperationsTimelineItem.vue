<script>
export default {
  name: 'OperationsTimelineItem',
  props: {
    operation: {
      type: Object,
      required: true,
    }
  },
  computed: {
    headerClass() {
      return {
        'INSERT': 'timeline-item__header--insert',
        'UPDATE': 'timeline-item__header--update',
        'DELETE': 'timeline-item__header--delete',
      }[this.operation.type]
    }
  },
}
</script>

<template>
  <div class="timeline-item">
    <div
      :class="headerClass"
      class="timeline-item__header"
    >
      <div class="type">
        {{ operation.type }}
      </div>

      <div class="date">
        <timeago
          :since="operation.timestamp"
          :auto-update="60"
        ></timeago>
      </div>
    </div>
    <div class="timeline-item__content">
      <div>
        <strong>{{ operation.schema }}:</strong>
        {{ operation.table }}
      </div>

      <i class="material-icons more-information">
        unfold_more
      </i>
    </div>
  </div>
</template>

<style scoped>
.timeline-item {
  position: relative;
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.06), 0 3px 6px rgba(0, 0, 0, 0.26);
}

.timeline-item__header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 4px 10px;
  background-color: #3787A8;
  color: #FFF;
}

.timeline-item__header--insert {
  background-color: #7DC6B9;
}

.timeline-item__header--update {
  background-color: #217092;
}

.timeline-item__header--delete {
  background-color: #D11345;
}

.timeline-item__content {
  display: grid;
  grid-template-columns: 1fr auto;
  justify-items: start;
  align-content: center;
  padding: 4px 5px 4px 10px;
  transition: .2s background-color;
  cursor: pointer;
}

.timeline-item__content:hover {
  background-color: #EBF3F6;
}

.more-information {
  grid-column: 2;
  justify-self: end;
  align-self: center;
  padding: 0;
  transition: .2s color;
}

.type {
  font-weight: 500;
}

.date {
  font-size: 12px;
  font-weight: 500;
}
</style>

