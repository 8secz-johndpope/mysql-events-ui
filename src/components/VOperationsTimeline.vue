<script>
import VOperationsTimelineItem from '@/components/VOperationsTimelineItem'

export default {
  name: 'OperationsTimeline',
  components: {
    VOperationsTimelineItem,
  },
  props: {
    operations: {
      type: Array,
      required: true,
    }
  }
}
</script>

<template>
  <div class="timeline">
    <div
      v-if="!operations.length"
      class="no-operations"
    >
      <i class="material-icons">
        timer_off
      </i>
      No operations on the moment
    </div>
    <VOperationsTimelineItem
      v-for="operation in operations"
      :key="operation.nextPosition"
      :operation="operation"
      class="timeline__item"
    />
  </div>
</template>

<style scoped>
.timeline {
  background-color: #fff;
  display: grid;
  grid-auto-flow: dense;
  grid-gap: 5px;
  grid-template-columns: repeat(auto-fit, minmax(180px, 1fr));
}

.no-operations {
  display: flex;
  align-items: center;
  justify-content: center;
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.06), 0 3px 6px rgba(0, 0, 0, 0.26);
  padding: 10px;
}

.no-operations > i {
  margin-right: 5px;
}
</style>

