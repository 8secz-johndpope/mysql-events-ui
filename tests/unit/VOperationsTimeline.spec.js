import { shallowMount } from '@vue/test-utils'
import VOperationsTimeline from '@/components/VOperationsTimeline'
import VOperationsTimelineItem from '@/components/VOperationsTimelineItem'
import socketFixture from './fixtures/socketServer'

describe('TheOperationsLog', () => {
  let props

  const build = () => {
    const wrapper = shallowMount(VOperationsTimeline, {
      propsData: props
    })

    return {
      wrapper,
      operationsTimelineItems: () => wrapper.findAll(VOperationsTimelineItem),
      noOperations: () => wrapper.find('.no-operations'),
    }
  }

  beforeEach(() => {
    props = {
      operations: []
    }
  })

  it('renders the component correctly', () => {
    // arranje
    const { wrapper } = build()

    // assert
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a list of timeline items', () => {
    // arranje
    props.operations = [
      socketFixture.response.INSERT,
      socketFixture.response.UPDATE,
      socketFixture.response.DELETE,
    ]
    const { operationsTimelineItems, noOperations } = build()
    const firstItem = operationsTimelineItems().at(0)

    // assert
    expect(operationsTimelineItems().length).toBe(props.operations.length)
    expect(firstItem.exists()).toBe(true)
    expect(firstItem.props().operation).toBe(props.operations[0])
    expect(noOperations().exists()).toBe(false)
  })

  it('updates the timeline list items', () => {
    // arranje
    const expectedOperations = [
      socketFixture.response.INSERT,
      socketFixture.response.UPDATE,
    ]

    props.operations = [
      socketFixture.response.INSERT,
      socketFixture.response.UPDATE,
      socketFixture.response.DELETE,
    ]
    const { wrapper, operationsTimelineItems } = build()
    expect(operationsTimelineItems().length).toBe(props.operations.length)

    // act
    wrapper.setProps({
      operations: expectedOperations
    })

    // assert
    expect(operationsTimelineItems().length).toBe(expectedOperations.length)
  })

  it('shows empty timeline when no operations are passed', () => {
    // arranje
    const { operationsTimelineItems, noOperations } = build()

    // assert
    expect(operationsTimelineItems().exists()).toBe(false)
    expect(noOperations().exists()).toBe(true)
    expect(noOperations().isVisible()).toBe(true)
  })
})