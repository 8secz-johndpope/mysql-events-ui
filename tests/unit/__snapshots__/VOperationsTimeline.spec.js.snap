// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VOperationsTimeline renders the component correctly 1`] = `
<div class="timeline">
  <div class="no-operations">
    <i class="material-icons">
      timer_off
    </i>
    No operations on the moment
  </div>
</div>
`;
